### ISSUE TITLE
Adding top-of-file metadata?

### ISSUE THREAD

**Issue by ndw** [bastibe/org-journal#182]
I've been creating daily journals for a while now, but in practice, I create a summary for every week. On the one hand, that's a good opportunity to synthesize the work of the week. On the other hand, I wouldn't have to cut and paste quite so much if I just kept a weekly journal.

So I'm thinking of switching to weekly.

But if I did that, I think I'd want to have a preamble:

```
#+TITLE: Some title, probably generated by a format-time-string or a function
#+OPTION: some option
#+HEADER: maybe other headers

* 6 January 2020
** 09:00 stuff I did
** 12:00 more stuff I did
* 7 January 2020
** 07:30 yet more stuff
...
```

On a casual inspection of the code, it didn't seem like this was configurable today. It's also not entirely clear if the extra headers would introduce any problems. I don't think they would, but...

Is this something you think could be made to work?

**Comment by bastibe** [comment]
Excellent idea! 

One would have to change `org-journal-new-entry` to check if the file is empty, and insert the contents of a new custom variable in that case.

Would you like to try implementing it yourself?

**Comment by cslux** [comment]
Hmm. Actually, this can already be done with `org-journal-data-format`, but the name of the variable is a bit misleading. 

``` emacs-lisp
(defun org-journal-date-format-func (time)
      "Custom function to insert journal date header,
  and some custom text on a newly created journal file."
      (concat (when (= (buffer-size) 0)
                (concat
                 (format-time-string
                  (pcase org-journal-file-type
                    (`daily "#+TITLE: Daily Journal, %A, %x")
                    (`weekly "#+TITLE: Weekly Journal No. %V, %Y")
                    (`monthly "#+TITLE: %B Journal, %Y")
                    (`yearly "#+TITLE: %Y Journal")))
                 "\n"))
              org-journal-date-prefix
              (format-time-string "%A, %x" time)))

(setq org-journal-date-format #'org-journal-date-format-func)
```

In your case.
``` emacs-lisp
(defun ndw/org-journal-monday (&optional time)
  "Return the TIME cast back to the previous Monday.
If TIME isn’t specified, the (current-time) is assumed.
If today is Monday, return today."
  (let* ((now    (if time time (current-time)))
         (dow    (string-to-number (format-time-string "%u" now)))
         (delta  (if (= dow 1)
                     0
                   (* (- dow 1) (* 24 3600)))))
    (seconds-to-time (- (time-to-seconds now) delta))))

(setq org-journal-date-format
      (lambda ()
        (concat (when (= (buffer-size) 0)
                  (let ((monday (ndw/org-journal-monday)))
                    (insert "#+TITLE: ")
                    (insert (format-time-string org-journal-date-format monday))
                    (insert "\n#+STARTUP: showeverything\n#+FILETAGS: :Journal:\n\n")))
                org-journal-date-prefix
                (format-time-string "%A, %x" time))))
```

**Comment by cslux** [comment]
@bastibe Separating the file header from the `org-journal-date-format` is a good idea. We had this discussion already once. In this change makes it clear what variable those what.

**Comment by bastibe** [comment]
Thank you very much @ndw for your contribution, and @cslux for integrating it!

**Comment by cslux** [comment]
Hey Mariano,

can you post your org-journal setup and, also the org-journal
version you use (MELPA version...)!

Thanks!

- Chris

**Comment by visika** [comment]
Sure!
Base config: the calendar works with this.
``` emacs-lisp
(use-package org-journal
      :straight t
      :ensure t
      :defer t
      :bind
      ("C-c n j" . org-journal-new-entry)
      :custom
      (org-journal-dir "~/org/")
  ;;    (org-journal-time-prefix "* ")
      (org-journal-file-format "%Y-%m-%d.org")
      (org-journal-enable-agenda-integration t)
```
Extension:
``` emacs-lisp
(defun org-journal-date-format-func (time)
      "Custom function to insert journal date header,
  and some custom text on a newly created journal file."
      (concat (when (= (buffer-size) 0)
                (concat
                 (format-time-string
                  (pcase org-journal-file-type
                    (`daily "#+TITLE: %d %B %Y")
                    (`weekly "#+TITLE: Weekly Journal No. %V, %Y")
                    (`monthly "#+TITLE: %B Journal, %Y")
                    (`yearly "#+TITLE: %Y Journal")))
                 "\n"))
              org-journal-date-prefix
              (format-time-string "%A, %x" time)))

(setq org-journal-date-format #'org-journal-date-format-func)
```
As for the version, my git folder is in sync with this repo with straight. Is this good practice or should I move org-journal to melpa?

**Comment by cslux** [comment]
Thanks, that helped me to find the issue quickly. Fixed via 271b99f.

> As for the version, my git folder is in sync with this repo with straight. Is this good practice or should I move org-journal to melpa?

Depends on the project IMO. But for org-journal it makes sense :-)

**Comment by visika** [comment]
Great work, thank you a lot!
